fix: buffered linear interpolator jitter and exposing more properties

NoelStephensUnity opened the pull request:
After doing some testing I noticed a few anomalies with the updated `BufferedLinearInterpolator<T>`. As it turns out, there are more than one approaches a user might need depending upon what they are trying to accomplish and each option really performs best when tweaking some properties within the `BufferedLinearInterpolator<T>` that were not completely exposed.

### Various options:
- You might want to enable, disable, or adjust the maximum interpolation time of the last smooth lerp phase for each interpolator type.
- You might want to manually adjust the tick latency offset (`NetworkTransform.InterpolationBufferTickOffset`) used depending upon context.
  - When using a client-server topology, you might want this to be set to something like 0 or 1 (depending upon interpolator).
    - You might want this value to be higher (like say +2-3 depending on latency and interpolation type) on non-authority instances where the authority is some other client and not the host/server.

### This PR renames `Lerp` to `LegacyLerp` and adds a 3rd  new `Lerp` interpolator type: 
####  Lerp (new)
Uses a 1 to 2 phase approach that lerps towards the target, lerps towards the next target (if one exists) ahead by 1 frame delta, blends the two results, and (optionally) smooth the final value.
- The first phase lerps towards the current tick state update being processed.
- The second phase (optional) performs a lerp smoothing where the current respective transform value is lerped towards the result of the first phase at a rate of 1.0f minus the max interpolation time.

## Changelog

- Added: `Lerp` interpolation type that still uses a lerp approach but uses the new buffer consumption logic.
- Added: Property to enable or disable lerp smoothing for position, rotation, and scale interpolators.
- Added: `NetworkTransform.InterpolationBufferTickOffset` static property to provide users with a way to increase or decrease the time marker where interpolators will pull state update from the queue.
- Fixed: Issue where the time delta that interpolators used would not be properly updated during multiple fixed update invocations within the same player loop frame.
- Changed: The original `Lerp` to be renamed to `LegacyLerp`.


## Testing and Documentation

- Includes integration test updates.
- Includes documentation for public API entry points.
- Includes updates to public documentation for `NetworkTransform`. (Will be adding to [PR-1443](https://github.com/Unity-Technologies/com.unity.multiplayer.docs/pull/1443))


<!--  Uncomment and mark items off with a * if this PR deprecates any API:
### Deprecated API
- [ ] An `[Obsolete]` attribute was added along with a `(RemovedAfter yyyy-mm-dd)` entry.
- [ ] An [api updater] was added.
- [ ] Deprecation of the API is explained in the CHANGELOG.
- [ ] The users can understand why this API was removed and what they should use instead.
-->

NoelStephensUnity committed:
wip

Getting closer to what I am hoping to achieve.

NoelStephensUnity committed:
update

Adding more properties for users to adjust as well as an additional interpolation type as there are various use cases where one model doesn't fit all genres.

NoelStephensUnity committed:
fix

Just wrapping some code that I will be removing anyway before marking this PR ready for review.
Adding some line breaks in new XML API documentation.

NoelStephensUnity committed:
update

Updating PositionMaxInterpolationTime, RotationMaxInterpolationTime, and ScaleMaxInterpolationTime to be set when `NetworkTransform.SetMaxInterpolationBound` is invoked.

NoelStephensUnity committed:
style

updating the xml api

NoelStephensUnity committed:
Merge branch 'develop-2.0.0' into fix/bufferedlinearinterpolator-jitter

NoelStephensUnity committed:
update

Defaulting the lerp smoothing to true (like it was originally when you couldn't enable/disable it).
Removing debug script from NetworkManager.

NoelStephensUnity committed:
style

Forgot to remove the components namespace when removing the debug script.

NoelStephensUnity committed:
Merge branch 'develop-2.0.0' into fix/bufferedlinearinterpolator-jitter

NoelStephensUnity committed:
fix

Only apply the InterpolationBufferTickOffset once.

NoelStephensUnity committed:
update

Bumping the default tick latency down to 1.

NoelStephensUnity committed:
fix

Fixing some additional issues and trying to find a reasonable balance.

NoelStephensUnity committed:
test

Updating NetworkTransformGeneral to test all interpolator types and adding a means to getting verbose debug logs,

NoelStephensUnity committed:
fix

Fixing a mistake on SD final lerp pass.

NoelStephensUnity committed:
test

making the added verbose log history internal to avoid PVP issue.

NoelStephensUnity committed:
fix

making sure the first entry is set to having reached its final point.

NoelStephensUnity committed:
test

Adding debug information to better understand the failure.

NoelStephensUnity committed:
update

Sigh...

NoelStephensUnity committed:
Update

Removing the teleport I put in last week.

NoelStephensUnity committed:
test

Dumb floating point precision.

NoelStephensUnity committed:
update

Minor adjustment UpdateInterpolation

NoelStephensUnity committed:
update

Removing commented out code.
NetworkTransform using the realtime provider time system to be more compatible with the integration test time travel system.

NoelStephensUnity committed:
test

providing a bit more time and reducing the target position to provide enough time to for interpolation.

NoelStephensUnity committed:
test

Removing some of the debug information.
Fixing whitespace issue.
Adding a smooth lerp and non-smooth lerp pass.

NoelStephensUnity committed:
text

Adding additional LerpExtrapolateBlend passes

NoelStephensUnity committed:
update

Minor adjustments after a long debug session...

NoelStephensUnity committed:
Merge branch 'develop-2.0.0' into fix/bufferedlinearinterpolator-jitter

NoelStephensUnity committed:
update

- Removed the lerp smoothing parameter and made it an internal bool (don't need to add to API and it should be set through the NetworkTransform anyway).
- Added additional BufferedLinearInterpolator debug code.
- Added additional PreUpdate pass for NetworkTransforms using Rigidbody motion in order to reset a fixed time delta.
- Added incemental fixed time delta when the FixedUpdate is invoked multiple times within a single frame.
- Added internal time sync counter.
- Added a "1/3rd" rule to smooth dampening and LerpExtrapolateBlend when smooth lerp is enabled in order to still be able to balance between smoothing the final lerped value and reaching the target (or getting much closer).

NoelStephensUnity committed:
style

removing the lerpSmoothing parameter from xml API documentation.

NoelStephensUnity committed:
style

One more removal of the same lerpSmoothing parameter in comments.

NoelStephensUnity committed:
Merge branch 'develop-2.0.0' into fix/bufferedlinearinterpolator-jitter

NoelStephensUnity committed:
Merge branch 'develop-2.0.0' into fix/bufferedlinearinterpolator-jitter

NoelStephensUnity committed:
style

updating xml api documentation to reflect the changes made.

NoelStephensUnity committed:
update

adding changelog entries.

NoelStephensUnity committed:
update

Some adjustments to  handle coming to a stop and to no longer clamp the timespan between state updates for LerpExtrapolateBlend and SmoothDampening.
Adding fixed delta to server time (if it is of any value then multiple fixed updates have run within a single frame).
Updated XML API documentation (again).

NoelStephensUnity committed:
style

XML API stuff.

NoelStephensUnity committed:
style

arrg... lists! PVP!

NoelStephensUnity committed:
style

2 more...evidently you can't use <list> within a <param>

NoelStephensUnity committed:
fix

Quaternion approximation calculation was wrong.

NoelStephensUnity committed:
update

Adding [MethodImpl(MethodImplOptions.AggressiveInlining)] to float, Vector3, and Quaternion LinearBufferInterpolator methods.

NoelStephensUnity committed:
fix

Updating the approximation approach to yield a more precise end result.

NoelStephensUnity committed:
fix

Don't try to convert  InterpolateState.Target if not assigned yet.

NoelStephensUnity committed:
update

Making smooth lerp a slightly faster for LEB and SD interpolation types.

EmandM commented on com.unity.netcode.gameobjects/Runtime/Timing/NetworkTimeSystem.cs:
```diff
@@ -257,7 +263,7 @@ public void Reset(double serverTimeSec, double rttSec)
             Advance(0);
         }
-
+        internal int SyncCount;
```
Out of interest, what is this count for?

NoelStephensUnity replied:
This is an internal counter that can be useful for debugging purposes and might be useful at a later time to provide users with additional debugging tools. Basically, this gets incremented every time a client receives a `TimeSyncMessage`... which can be checked for changes...which then provides one with the ability to determine the actual local time frequency of receiving and processing time sync messages. 
_(I use it in the added debug stuff within `NetworkTransform` to know on what time synch message count a state update was received as well to be able to check if there is any variance in the time to target for a state update that occurs within the same approximate time frame that a time synchronization occurs)_

EmandM commented on com.unity.netcode.gameobjects/Runtime/Timing/NetworkTimeSystem.cs:
```diff
@@ -82,10 +82,16 @@ public class NetworkTimeSystem
+        /// - <see cref="Components.NetworkTransform.GetTickLatencyInSeconds"/> <br />
         /// </remarks>
-        public int TickLatency = 2;
+        public int TickLatency = 1;
```
Will there be any side effects from changing the default of this value?

NoelStephensUnity replied:
Yes. This was originally created in [PR-3337](https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/pull/3337/files#diff-66360097061d73228ae288f38b02f6fed365b52920b5ae021cebb7c5da837e1a) which was not a reasonable value to start with. Tick latency will be adjusted based on the latency from the server (NGO or Rust) as it updates and synchronizes time. However, the host/server should really stay at a tick latency of 1 until we can get clients sending a response back to the server when it gets a network time synchronization so the server can make adjustments dynamically (per client) to this value... until then it should be 1 _(I can explain in better detail in our next 1:1)_

EmandM commented on com.unity.netcode.gameobjects/Runtime/Components/Interpolator/BufferedLinearInterpolator.cs:
```diff
@@ -350,23 +428,63 @@ private void TryConsumeFromBuffer(double renderTime, float minDeltaTime, float m
         /// <param name="tickLatencyAsTime">The tick latency in relative local time.</param>
         /// <param name="minDeltaTime">The minimum time delta between the current and target value.</param>
         /// <param name="maxDeltaTime">The maximum time delta between the current and target value.</param>
+        /// <param name="isLerpAndExtrapolate">Determines whether to use smooth dampening or extrapolation.</param>
```
nit: this param is a bit unclear
```suggestion
        /// <param name="isLerpAndExtrapolate">True when using <see cref="NetworkTransform.InterpolationTypes.Lerp"/>; otherwise false.</param>
```

NoelStephensUnity replied:
This is exactly as it should be.  👌 
`NetworkTransform.InterpolationTypes.Lerp` uses the `BufferedLinearInterpolator<T>.Update(float deltaTime, double renderTime, double serverTime)` method and `NetworkTransform.InterpolationTypes.LerpExtrapolateBlend` and `NetworkTransform.InterpolationTypes.SmoothDampening` use the `BufferedLinearInterpolator<T>.Update(float deltaTime, double tickLatencyAsTime, double minDeltaTime, double maxDeltaTime, bool isLerpAndExtrapolate)` method when being updated by `NetworkTransform`.

EmandM commented on com.unity.netcode.gameobjects/Runtime/Components/Interpolator/BufferedLinearInterpolator.cs:
```diff
@@ -220,6 +249,8 @@ public void Reset(T currentValue)
         /// </summary>
         private protected bool m_IsAngularValue;
 
+        private bool m_WasPredictedLerp;
```
This doesn't seem to be used

NoelStephensUnity replied:
Nice catch! 
Thank you for the attention to detail... I need a second set of eyes and keep up the questions... I need everything with this update scrutinized over for sure. 👍

EmandM commented on com.unity.netcode.gameobjects/Runtime/Components/NetworkTransform.cs:
```diff
@@ -3858,59 +4078,165 @@ private void UpdateInterpolation()
                 {
-                    m_PositionInterpolator.Update(cachedDeltaTime, tickLatencyAsTime, minDeltaTime, maxDeltaTime);
+                    m_PositionInterpolator.Update(cachedDeltaTime, tickLatencyAsTime, minDeltaTime, maxDeltaTime,
+                        PositionInterpolationType == InterpolationTypes.LerpExtrapolateBlend);
```
Here this parameter is true when `PositionInterpolationType == InterpolationTypes.LerpExtrapolateBlend` but where you're using this variable the code comments indicate that this parameter refers to `PositionInterpolationType == InterpolationTypes.Lerp`

NoelStephensUnity replied:
This bool parameter is used to determine if the update is to use LerpExtrapolateBlend or if not then it uses SmoothDampening. Does the parameter description need to be updated for better clarity?
`/// <param name="isLerpAndExtrapolate">Determines whether to use smooth dampening or extrapolation.</param>`

EmandM commented on com.unity.netcode.gameobjects/Runtime/Components/Interpolator/BufferedLinearInterpolator.cs:
```diff
@@ -350,23 +428,63 @@ private void TryConsumeFromBuffer(double renderTime, float minDeltaTime, float m
+                    InterpolateState.AddDeltaTime(deltaTime);
+                    var targetValue = InterpolateState.CurrentValue;
+                    // SmoothDampen or LerpExtrapolateBlend
+                    if (!isLerpAndExtrapolate)
```
Where this function is called `isLerpAndExtrapolate` is being set as `PositionInterpolationType == InterpolationTypes.LerpExtrapolateBlend`. This code comment doesn't seem to match.

NoelStephensUnity replied:
It evaluates to a Boolean where if using `InterpolationTypes.LerpExtrapolateBlend` it Lerps then Lerps unclamped (extrapolates) where if it is not then it does the two pass smooth dampening. It is sort of inverted logic...
Use the smooth dampening if we are not doing `LerpExtrapolateBlend` is what that turns out to be logically.

EmandM replied:
I missed that there were two `Update` functions! I follow it now

EmandM reviewed (commented):
Not sure I understand all of it, but here is an initial pass review :)

NoelStephensUnity committed:
update

Exposing internal access to position and rotation interpolators for future tests and tools.
Noticed the slerp for full precision rotation was only being applied when using Lerp, now it is for all interpolation types.
Making the time to approximately reach adjustment be the default when there is no pending item to be consumed and slightly sooner when there is.
Removing the adjustment to the slerp smoothing as this should be a consistantly applied value for all interpolation types.

NoelStephensUnity committed:
test

Renaming NetworkTransformMultipleChangesOverTime to just MultipleChangesOverTime (no need for the NetworkTransform part as it is already part of the NetworkTransformTests).
Turning off Lerp smoothing for MultipleChangesOverTime and added lengthy comment as to why.
Fixed the initial check for the first state update in MultipleChangesOverTime  due to a change in the Interpolation value. It was not properly checking both pushed and updated values and should only be checked when interpolation is turned off as the default is enabled and when running MultipleChangesOverTime  for 3 axis it remains enabled (i.e. there will be no state update for that as nothing changed).

NoelStephensUnity committed:
style

removing unused property.

NoelStephensUnity committed:
Merge branch 'develop-2.0.0' into fix/bufferedlinearinterpolator-jitter

NoelStephensUnity committed:
update

Some clean up, renaming, and removing the change in the original BufferedLinearInterpolator<T>.ResetTo method.

NoelStephensUnity committed:
Merge branch 'develop-2.0.0' into fix/bufferedlinearinterpolator-jitter

EmandM reviewed (approved):
Ok! I get it! There's one `Update()` for Lerp, and a separate one for Smooth Dampening and LerpExtrapolateBlend! Ok, this looks good!

EmandM committed:
Merge branch 'develop-2.0.0' into fix/bufferedlinearinterpolator-jitter

NoelStephensUnity committed:
update

Renamed LerpExtrapolateBlend to LerpAhead.
Found some issues with the original lerp's buffer consumption where it was always calculating 2x the time to target.
Assuring that the original lerp logic isn't executed upon arriving to the final destination.
Increasing the k_AproximatePrecision...preceision value to be one one hundred thousandth.
Removed some debug code from NetworkTransformBase.

NoelStephensUnity committed:
Merge branch 'develop-2.0.0' into fix/bufferedlinearinterpolator-jitter

NoelStephensUnity committed:
style

Adjusting changelog entry to replace LerpExtrapolateBlend with LerpAhead.
Removing additional remarks tag.

NoelStephensUnity committed:
Merge branch 'fix/bufferedlinearinterpolator-jitter' of https://github.com/Unity-Technologies/com.unity.netcode.gameobjects into fix/bufferedlinearinterpolator-jitter

NoelStephensUnity committed:
test

Updating the test project test for change in LerpAhead.

NoelStephensUnity committed:
update

Reverting the change to fix that makes the original lerp properly calculate the time to target as opposed to always calculating 2x the time to target because all of the unit interpolator tests are specifically written for this specific flaw.

NoelStephensUnity committed:
style

Whitespace fixes.

NoelStephensUnity committed:
update

Still had one more line of code needed to go back to original lerp buffer consumption where the time to target is always 2x the actual time to target (it ends up consuming 2 state updates for each non-authority update).

Making sure some of the debugging BLI accessor methods in NetworkTransform are still accessible for internal testing purposes.

NoelStephensUnity committed:
refactor

After testing for quite some time I came to the conclusion that we needed a "new" lerp and should mark the original lerp as the "legacy" lerp while keeping the added option to have the smooth dampening.

Removed the additional phases from the "new" lerp and smooth dampening while also providing users with a more direct and constant maximum interpolation setting that is not directly tied to the frame time delta.

This yields a wider range of fine tuning and will be easier to handle adjusting during runtime.

The LinearBufferInterpolator RNSM component hooks are in place so we can provide users with a tool to be able to determine what is happening with the LinearBufferInterpolator... it will be released in the examples section once ready to be released.

Adjusted API XML documentation and all places that referenced the other enum names changed.

NoelStephensUnity committed:
update

updating changelog entry.
